### PR TITLE
#238 useDocument naming and tests

### LINKED ISSUE #238: `useDocument` naming is confusing
As per : https://github.com/openmsupply/openmsupply-client/pull/228#pullrequestreview-775707200

The hook `useDocument` has a bunch of types and the naming makes it more difficult than it should be to understand what is going on

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #238 

- Did some renaming. Hope it's clearer
- Also added some tests for `useDocument` and the `DetailView` reducer.

### CODE REVIEW

File: packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    // Create some server data which is the same except every line has 99 quantity.
+    // Then after merging, every line should have 99 quantity.
+    const dataLines = lines.map(line => ({ ...line, quantity: 99 }));
+    const data: Transaction = { ...placeholderTransaction, lines: dataLines };
+
+    const reducerResult = reducer(data, null)(state, DocumentAction.merge());
+
+    // Check for any lines that don't have a quantity of 99. If there are any, the merge was wrong.
+    expect(
+      reducerResult.draft.lines.filter(({ quantity }) => quantity !== 99).length
```
Comment (joshxg) [inline packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts:203]: Probably a more comprehensive 'merging' might be needed. But this test I think is good for showing the feature. For example, we might not want the quantity to be overridden but instead have a `clientSideEnteredValueForQuantityHere` and a `quantity` to show 'you entered X but it wasn't correctly updated on the server for Y reason, so we've changed it to  Y' kind of thing.

Additionally need to handle other people deleting things etc (ideally not just removing the line a user might be editing at the time..?)

File: packages/common/src/hooks/useDocument/types.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+export enum DocumentActionType {
+  Init = 'Draft/init',
+  Merge = 'Draft/merge',
+}
+
+export type DefaultDocumentAction =
+  | { type: DocumentActionType.Init }
+  | { type: DocumentActionType.Merge };
+
+export type DocumentActionSet<T> = T | DefaultDocumentAction;
```
Comment (joshxg) [inline packages/common/src/hooks/useDocument/types.ts:29]: Renamed things here. Hope it's clearerer?

File: packages/common/src/hooks/useDocument/useDocument.test.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  TestingProvider,
+  TestingRouter,
+} from '@openmsupply-client/common';
+import { waitFor } from '@testing-library/dom';
+import { useDocument } from './useDocument';
+import { Api } from './types';
+import { renderHook } from '@testing-library/react-hooks';
+import { Route } from 'react-router';
+
+describe('useDocument', () => {
```
Comment (joshxg) [inline packages/common/src/hooks/useDocument/useDocument.test.tsx:13]: these tests are pretty lame but I think show a fine example of a simplified version of using the hook

File: packages/common/src/hooks/useDocument/useDocument.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import { useEffect, useReducer, Dispatch, useRef } from 'react';
+import { useNavigate } from 'react-router-dom';
+import { useMutation, useQuery, useQueryClient } from 'react-query';
+
+import {
```
Comment (joshxg) [inline packages/common/src/hooks/useDocument/useDocument.ts:5]: I renamed things here and it did a full diff

File: packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+    expect(reducerResult.draft.lines).toEqual(
+      [
+        expect.objectContaining({ id: '1' }),
+        expect.objectContaining({ id: '1' }),
+        expect.objectContaining({ id: '2' }),
+        expect.objectContaining({ id: '3' }),
+        expect.objectContaining({ id: '4' }),
+        expect.objectContaining({ id: '5' }),
+      ].reverse()
```
Comment (mark-prins) [inline packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts:134]: sneaky `reverse`!! I didn't notice that at first

File: packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts (hunk 1 above)
Comment (mark-prins) [reply packages/transactions/src/OutboundShipment/DetailView/reducer.test.ts:203]: yeah, would be neat to tell people that someone else deleted a line

**Review by mark-prins — COMMENTED** [review]
👍  woo! that looks great. tests are passing, and clearer to read through

**Review by mark-prins — APPROVED** [review]
hm. I commented and didn't approve? oops.

**Pull request merged**